SWLOR.Game.Server/Native/ResolveAttackRoll.cs

```diff
@@ -24,7 +28,366 @@ public static void RegisterHook()
+            // Note: this always returns object invalid for NPCs (2130706432) as their actions aren't represented the same way.
+            uint oidTarget = defender.m_pActionQueue.GetItem(0).oidTarget;
+
+            if (oidTarget == 2130706432) // OBJECT_INVALID
+            {
+                oidTarget = (uint) defender.m_ScriptVars.GetInt(new CExoString("I_LAST_ATTACKED"));
+            }
+
+            // If this is an NPC attacking, Store the attack on the NPC. 
+            if (attacker.m_pActionQueue.GetItem(0).oidTarget == 2130706432)
+            {
+                Log.Write(LogGroup.Attack, "NPC attacking - storing target "+ defender.m_idSelf.ToString());
+                attacker.m_ScriptVars.SetInt(new CExoString("I_LAST_ATTACKED"), (int)defender.m_idSelf);
+            }
+
+            // oidTarget will be 0 for a newly spawned NPC who hasn't been attacked yet.  Don't let them get taken by surprise in round 1. 
+            if (oidTarget != 0 && oidTarget != attacker.m_idSelf)
+            {
+                Log.Write(LogGroup.Attack, "Defender current target ("+oidTarget.ToString() +") is not attacker ("+attacker.m_idSelf.ToString()+"). Assign circumstance bonus");
+                modifiers += 5;
+            }
+
+            // Weapon focus feats.
+            modifiers += 5 * HasWeaponFocus(attacker, weapon);
+
+            // Range bonuses and penalties.
+            if (attackType == (uint)AttackType.Ranged || attackType == (uint)AttackType.Spirit)
+            {
+                Vector attackerPos = attacker.m_vPosition;
+                Vector defenderPos = defender.m_vPosition;
+
+                // Note - calculating distance solely via X/Y co-ordinates.  NWN doesn't have a true Z.
+                double range = Math.Pow(Math.Pow((attackerPos.x - defenderPos.x), 2) + Math.Pow((attackerPos.y - defenderPos.y),2), 0.5);
+                     
+                Log.Write(LogGroup.Attack, "Ranged attack at range " + range.ToString());
+                if (range < 5.0f)
+                {
+                    // Force powers or point blank shot feat make close range an advantage.
+                    if (attackType == 3 || attacker.m_pStats.HasFeat((ushort)FeatType.PointBlankShot) == 1)
+                    {
+                        modifiers += 5;
+                    }
+                    else
+                    {
+                        modifiers -= 20;
+                    }
+                }
+                else if (range > 40.0f)
+                {
+                    modifiers -= 20;
+                }
+                else if (range > 30.0f)
+                {
+                    modifiers -= 10;
+                }
+                else if (range > 20.0f)
+                {
+                    modifiers = -5;
+                }
+                
+            }
+
+            // Attacking from behind.  Does not apply to Force attacks.
```

Comment:
I really like this idea but I'm wondering if it'd be better served as a perk instead of being freely available.

Reply:
Could have a perk for extra damage?   
Making this globally available is mainly about giving parties ganging up on someone an advantage - whether that's PCs or NPCs.  It means that even if a single Mynock can't really touch you, a swarm of them will have a decent shot at getting through.  Equally, I don't feel strongly about this if you want to lock it behind a perk.

Reply:
Yeah let's keep it. We can investigate adding a perk bonus at a later date.